File: apps/mobile/courier-android/src/main/kotlin/com/fleetfeast/mobile/courier/CourierBackendClient.kt
package com.fleetfeast.mobile.courier

import java.net.HttpURLConnection
import java.net.URI
import java.net.URLEncoder
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

data class HttpTransportRequest(
  val method: String,
  val url: String,
  val headers: Map<String, String> = emptyMap(),
  val body: String? = null,
)

data class HttpTransportResponse(
  val statusCode: Int,
  val body: String? = null,
)

interface HttpTransport {
  fun execute(request: HttpTransportRequest): HttpTransportResponse
}

class UrlConnectionHttpTransport : HttpTransport {
  override fun execute(request: HttpTransportRequest): HttpTransportResponse {
    val connection = URI.create(request.url).toURL().openConnection() as HttpURLConnection
    connection.requestMethod = request.method
    connection.connectTimeout = 10_000
    connection.readTimeout = 10_000

    request.headers.forEach { (name, value) ->
      connection.setRequestProperty(name, value)
    }

    if (request.body != null) {
      connection.doOutput = true
      connection.outputStream.use { output ->
        output.write(request.body.encodeToByteArray())
      }
    }

    val status = connection.responseCode
    val stream = if (status in 200..299) connection.inputStream else connection.errorStream
    val body = stream?.use { input ->
      input.readBytes().decodeToString()
    }

    connection.disconnect()
    return HttpTransportResponse(statusCode = status, body = body)
  }
}

@Serializable
data class CourierJob(
  val jobId: String,
  val orderId: String,
  val status: String,
)

@Serializable
data class FeatureFlagSnapshot(
  val flags: Map<String, Boolean>,
  val ttlSeconds: Int,
  val generatedAtEpochMillis: Long,
)

@Serializable
data class AppSession(
  val sessionId: String,
  val userId: String,
  val role: String,
  val persona: String,
  val traceId: String,
  val refreshTokenId: String,
  val issuedAt: String,
  val expiresAt: String,
)

@Serializable
data class AppSessionTokenPair(
  val tokenType: String,
  val accessToken: String,
  val refreshToken: String,
  val expiresInSeconds: Int,
  val refreshExpiresInSeconds: Int,
  val refreshExpiresAt: String,
)

@Serializable
data class SessionExchangeResponse(
  val session: AppSession,
  val tokenPair: AppSessionTokenPair,
)

@Serializable
private data class CourierJobsPayload(
  val jobs: List<CourierJob>,
)

@Serializable
private data class CourierSessionExchangeRequest(
  val oidcToken: String,
  val traceId: String,
  val deviceId: String,
)

@Serializable
private data class CourierSessionRefreshRequest(
  val refreshToken: String,
  val traceId: String,
  val deviceId: String,
)

class CourierBackendClient(
  private val baseUrl: String,
  private val transport: HttpTransport = UrlConnectionHttpTransport(),
) {
  private val json = Json {
    ignoreUnknownKeys = true
    explicitNulls = false
  }

  init {
    require(baseUrl.startsWith("http://") || baseUrl.startsWith("https://")) {
      "baseUrl must be a valid http/https URL"
    }
  }

  fun fetchAvailableJobs(): List<CourierJob> {
    val response = executeGet("/app/v1/courier/jobs/available")
    return json.decodeFromString<CourierJobsPayload>(response.body ?: "{}").jobs
  }

  fun fetchFeatureFlags(
    userId: String,
    role: String,
    tenantId: String?,
  ): FeatureFlagSnapshot {
    val query = linkedMapOf(
      "userId" to userId,
      "role" to role,
    )
    if (tenantId != null) {
      query["tenantId"] = tenantId
    }

    val response = executeGet(
      path = "/app/v1/courier/feature-flags",
      query = query,
    )
    return json.decodeFromString(response.body ?: "{}")
  }

  fun exchangeSession(
    oidcToken: String,
    traceId: String,
    deviceId: String,
  ): SessionExchangeResponse {
    val requestBody = CourierSessionExchangeRequest(
      oidcToken = oidcToken,
      traceId = traceId,
      deviceId = deviceId,
    )
    val response = executePost(
      path = "/app/v1/courier/session/exchange",
      body = json.encodeToString(requestBody),
    )
    return json.decodeFromString(response.body ?: "{}")
  }

  fun refreshSession(
    refreshToken: String,
    traceId: String,
    deviceId: String,
  ): SessionExchangeResponse {
    val requestBody = CourierSessionRefreshRequest(
      refreshToken = refreshToken,
      traceId = traceId,
      deviceId = deviceId,
    )
    val response = executePost(
      path = "/app/v1/courier/session/refresh",
      body = json.encodeToString(requestBody),
    )
    return json.decodeFromString(response.body ?: "{}")
  }

  private fun executeGet(path: String, query: Map<String, String> = emptyMap()): HttpTransportResponse {
    val response = transport.execute(
      HttpTransportRequest(
        method = "GET",
        url = buildUrl(path, query),
      ),
    )

    if (response.statusCode !in 200..299) {
      throw IllegalStateException("Request failed with status ${response.statusCode} for $path")
    }

    return response
  }

  private fun executePost(path: String, body: String): HttpTransportResponse {
    val response = transport.execute(
      HttpTransportRequest(
        method = "POST",
        url = buildUrl(path, emptyMap()),
        headers = mapOf("content-type" to "application/json"),
        body = body,
      ),
    )

    if (response.statusCode !in 200..299) {
      throw IllegalStateException("Request failed with status ${response.statusCode} for $path")
    }

    return response
  }

  private fun buildUrl(path: String, query: Map<String, String>): String {
    val normalizedBase = baseUrl.removeSuffix("/")
    if (query.isEmpty()) {
      return "$normalizedBase$path"
    }

    val queryString = query.entries.joinToString("&") { (key, value) ->
      "${encodeQuery(key)}=${encodeQuery(value)}"
    }
    return "$normalizedBase$path?$queryString"
  }

  private fun encodeQuery(value: String): String {
    return URLEncoder.encode(value, Charsets.UTF_8).replace("+", "%20")
  }
}

File: apps/mobile/courier-ios/Sources/CourierIOSShell/CourierBackendClient.swift
import Foundation

public protocol HTTPClient {
    func request(_ request: URLRequest) async throws -> (Data, HTTPURLResponse)
}

public struct URLSessionHTTPClient: HTTPClient {
    private let session: URLSession

    public init(session: URLSession = .shared) {
        self.session = session
    }

    public func request(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw CourierBackendClientError.invalidResponse
        }

        return (data, httpResponse)
    }
}

public struct CourierJob: Decodable {
    public let jobId: String
    public let orderId: String
    public let status: String
}

public struct FeatureFlagSnapshot: Decodable {
    public let flags: [String: Bool]
    public let ttlSeconds: Int
    public let generatedAtEpochMillis: Int64
}

public enum CourierBackendClientError: Error {
    case invalidResponse
    case requestFailed(statusCode: Int)
}

private struct CourierJobsEnvelope: Decodable {
    let jobs: [CourierJob]
}

public struct CourierBackendClient {
    private let baseURL: URL
    private let httpClient: HTTPClient
    private let decoder = JSONDecoder()

    public init(baseURL: URL, httpClient: HTTPClient = URLSessionHTTPClient()) {
        self.baseURL = baseURL
        self.httpClient = httpClient
    }

    public func fetchAvailableJobs() async throws -> [CourierJob] {
        let data = try await performGET(path: "/app/v1/courier/jobs/available")
        return try decoder.decode(CourierJobsEnvelope.self, from: data).jobs
    }

    public func fetchFeatureFlags(
        userId: String,
        role: String,
        tenantId: String?
    ) async throws -> FeatureFlagSnapshot {
        var queryItems = [
            URLQueryItem(name: "userId", value: userId),
            URLQueryItem(name: "role", value: role),
        ]
        if let tenantId {
            queryItems.append(URLQueryItem(name: "tenantId", value: tenantId))
        }

        let data = try await performGET(
            path: "/app/v1/courier/feature-flags",
            queryItems: queryItems
        )
        return try decoder.decode(FeatureFlagSnapshot.self, from: data)
    }

    private func performGET(
        path: String,
        queryItems: [URLQueryItem] = []
    ) async throws -> Data {
        let base = baseURL.absoluteString.replacingOccurrences(of: "/$", with: "", options: .regularExpression)
        var components = URLComponents(string: "\(base)\(path)")
        components?.queryItems = queryItems.isEmpty ? nil : queryItems

        guard let finalURL = components?.url else {
            throw CourierBackendClientError.invalidResponse
        }

        var request = URLRequest(url: finalURL)
        request.httpMethod = "GET"

        let (data, response) = try await httpClient.request(request)
        guard (200...299).contains(response.statusCode) else {
            throw CourierBackendClientError.requestFailed(statusCode: response.statusCode)
        }

        return data
    }
}
